### PR TITLE
Add test for global Nosé-Hoover thermostat + refactor

### PR DESCRIPTION
**Pull request by danielhollas**

### CODE REVIEW

File: src/nosehoover.F90 [hunk 1]
```diff
@@ -67,95 +93,202 @@ subroutine calc_nhcham()
...
+         if (iknow /= 1) error = .true.
+      end if
+
+      if (error) then
+         call fatal_error(__FILE__, __LINE__, &
+            'Invalid NHC thermostat parameter')
+      end if
+   end subroutine check_nhc_parameters
 
    subroutine nhc_init()
```
Comment (danielhollas) [inline src/nosehoover.F90:174]: This init function was quite messy so I've broken in down to several pieces in individual functions. I also moved keyword sanity checks related to NHC here from init.F90.

File: src/nosehoover.F90 [hunk 2]
```diff
@@ -67,95 +93,202 @@ subroutine calc_nhcham()
...
       if (ipimd == 1 .and. inormalmodes /= 1) then
 
-         do iw = 1, nwalk
-            Qm(iw) = 1 / (TEMP * NWALK)
+         ! TODO: Is this optimal both with/wo staging transform?
+         Qm(1) = nhc_mass ! see tuckermann,stat.mech.
+         do iw = 2, nwalk
+            Qm(iw) = 1 / (temp * nwalk)
          end do
-         if (tau0 > 0) Qm(1) = ams ! see tuckermann,stat.mech.
```
Comment (danielhollas) [inline src/nosehoover.F90:133]: Here I made an actual functional change for PIMD with staging transformation. Previously, if `tau0` was not set in the input file, the NHC mass for the staging centroid would be the same as for the other beads. Not sure whether this has real consequence, I'll do some testing later. Need to read the Tuckermann again to better understand where these formulas are coming from.

**Review by suchanj — APPROVED** [review]

File: src/nosehoover.F90 [hunk 3]
```diff
@@ -67,95 +93,202 @@ subroutine calc_nhcham()
...
-      allocate (ms(nmolt, nchain))
-      allocate (ran(natom * 3))
+      call check_nhc_parameters()
 
-      ! SETTING THERMOSTAT MASSES
-      ams = ams * amu
-      if (ams < 0 .and. tau0 < 0) then
+      call set_yoshida_weights(nyosh)
+
+      ! Not sure if we should have this default value
```
[the remark this answers is not in the document]
Comment (danielhollas) [reply src/nosehoover.F90:188]: That is a very good point, I've been worried about that for some time, but didn't know how to approach it.
I would suggest for now to remove the default value for classical MD and require it to be set explicitly in the input file, which should at least alert the users for it's importance a bit. What do you think? We could maybe keep the default value for PIMD.

Writing something up would be great, perhaps I could do a seminar in our group here and than share the slides.

[the remark this answers is not in the document]
Comment (danielhollas) [reply src/nosehoover.F90:183]: @suchanj I've opened #129 and #130 to investigate optimal setting for NHC thermostat in the context of classical MD.

**Pull request merged**